### PR TITLE
Use Typescript

### PR DESCRIPTION
**Pull request by MyIgel**
## Description

This replaces most .js files with their typescript equivalent.

## Motivation and Context

Having no clue what something does is bad, this fixes it b having types and thus at least some clues :tada: 

## How Has This Been Tested?

On my laptop, technically should produce the same JS as before

## Screenshots/links:
n/a

## Checklist:

- [x] My code follows the code style of this project. (CI will test it anyway and also needs approval)
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by MyIgel** [comment]
The only anoyance is the fact that `forcegraph.js`'s `transformPosition` sets the `x`, `y`, & `k` parameters which are, according to the typescript doc, constants and thus triggers ide warnings.
My attempts at fixing that where not successful, the following change produces random jumps when klicking around a node in graph view after highlighting it:

```js
transform = new ZoomTransform(p.k, p.x, p.y);
draw.setTransform(transform);
redraw();
```

**Comment by MyIgel** [comment]
The only anoyance is the fact that `forcegraph.js`'s `transformPosition` sets the `x`, `y`, & `k` parameters which are, according to the typescript doc, constants and thus triggers ide warnings.
My attempts at fixing that where not successful, the following change produces random jumps when klicking around a node in graph view after highlighting it:

```js
transform = new ZoomTransform(p.k, p.x, p.y);
draw.setTransform(transform);
redraw();
```

Maybe @weeman1337 has an idea here?

File: lib/about.ts [hunk 1]
```diff
@@ -1,5 +1,8 @@
-export const About = function (picturesSource, picturesLicense) {
-  this.render = function render(d) {
+import { CanRender } from "./container";
+
+export const About = function (picturesSource: string, picturesLicense: string): CanRender {
+  function render(d: HTMLElement) {
+    let _ = window._;
```
Comment (weeman1337) [inline lib/about.ts:5]: This looks like a global function → I would suggest to export it from a module and add type specs.

File: lib/about.ts [hunk 2]
```diff
@@ -58,5 +61,9 @@ export const About = function (picturesSource, picturesLicense) {
       "<p>The source code is available at " +
       '<a href="https://github.com/freifunk/meshviewer">' +
       "https://github.com/freifunk/meshviewer</a>.</p>";
+  }
+
+  return {
+    render: render,
```
Comment (weeman1337) [inline lib/about.ts:67]:
```suggestion
    render,
```

File: lib/container.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export interface CanAdd {
+  add: (element: CanRender) => any;
+}
+
+export const Container = function (tag?: string): CanRender & CanAdd {
+  if (!tag) {
+    tag = "div";
+  }
+
+  let self = {
```
Comment (weeman1337) [inline lib/container.ts:14]: If a variable is not reassigned, it should be declared as `const`
(counts for other places as well)

File: lib/filters/hostname.ts [hunk 4]
```diff
@@ -1,7 +1,11 @@
...
-  var refreshFunctions = [];
-  var timer;
-  var input = document.createElement("input");
+import { Node } from "../utils/node";
+import { CanRender } from "../container";
+import { Filter } from "../datadistributor";
+
+export const HostnameFilter = function (): CanRender & Filter {
+  let refreshFunctions: (() => any)[] = [];
+  let timer: NodeJS.Timeout;
```
Comment (weeman1337) [inline lib/filters/hostname.ts:7]: Technically `NodeJS` doesn't exist in the browser.  
You can use this to get the correct type in any case:

```suggestion
  let timer: ReturnType<typeof setTimeout>;
```

File: lib/forcegraph.ts [hunk 5]
```diff
@@ -42,64 +54,69 @@ export const ForceGraph = function (linkScale, sidebar) {
     draw.setMaxArea(canvas.width, canvas.height);
   }
 
-  function transformPosition(p) {
+  function transformPosition(p: { k: number; x: number; y: number }) {
+    // @ts-ignore
```
Comment (weeman1337) [inline lib/forcegraph.ts:58]: To solve this ignore transform should be replaced with a new instance 

```
transform = new d3Zoom.ZoomTransform(p.k, p.x, p.y);
```

File: lib/forcegraph.ts [hunk 6]
```diff
@@ -42,64 +54,69 @@ export const ForceGraph = function (linkScale, sidebar) {
...
 
-    var e = transform.invert([event.clientX, event.clientY]);
-    var node = force.find(e[0], e[1], NODE_RADIUS_SELECT);
+    let click = transform.invert([event.clientX, event.clientY]);
+    let point = { x: click[0], y: click[1] };
+    let node = force.find(point.x, point.y, NODE_RADIUS_SELECT);
+    let router = window.router;
 
     if (node !== undefined) {
+      // @ts-ignore
```
Comment (weeman1337) [inline lib/forcegraph.ts:111]: Is `node` actually something different here?

File: lib/gui.ts [hunk 7]
```diff
@@ -46,39 +52,40 @@ export const Gui = function (language) {
     content = null;
   }
 
-  function addContent(Content) {
+  function addContent(mapViewComponent: typeof Map | typeof ForceGraph) {
     removeContent();
 
-    content = new Content(linkScale, sidebar, buttons);
+    // @ts-ignore
```
Comment (weeman1337) [inline lib/gui.ts:58]: This can probably be removed.

File: lib/infobox/main.ts [hunk 8]
```diff
@@ -26,9 +36,12 @@ export const Main = function (sidebar, linkScale) {
 
     el.scrollIntoView(false);
     el.classList.add("infobox");
+    // @ts-ignore
```
Comment (weeman1337) [inline lib/infobox/main.ts:39]: If `destroy()` should be kept with the element, you could use an extended element with the `destroy()` function defined.

File: lib/proportions.ts [hunk 9]
```diff
@@ -226,15 +235,17 @@ export const Proportions = function (filterManager) {
...
 
-  self.renderSingle = function renderSingle(el, heading, table) {
+  self.renderSingle = function renderSingle(el: HTMLElement, heading: string, table: HTMLTableElement) {
     if (table.children.length > 0) {
-      var h2 = document.createElement("h2");
+      let h2 = document.createElement("h2");
       h2.classList.add("proportion-header");
       h2.textContent = _.t(heading);
       h2.onclick = function onclick() {
+        // @ts-ignore
```
Comment (weeman1337) [inline lib/proportions.ts:244]: `renderSingle()` should return an extended table with the `elm` prop.

**Review by weeman1337 — CHANGES_REQUESTED** [review]
:partying_face: thanks for migrating the project to TypeScript. Left some notes.

Also ESLint explodes, because it cannot parse TypeScript:
- `npm i --save-dev @typescript-eslint/parser`
- Add `"parser": "@typescript-eslint/parser"` to `.eslintrc.json`

File: lib/container.ts (hunk 3 above)
Comment (MyIgel) [reply lib/container.ts:14]: I wanted to change as little as needed here to not make the changes more confusing than they  already are but can change it at least for the self's

File: lib/forcegraph.ts (hunk 5 above)
Comment (MyIgel) [reply lib/forcegraph.ts:58]: I would love doing that but that breaks it even more, see the comments above.
Can you eventually show me exactly how you mean that at some point?

File: lib/forcegraph.ts (hunk 6 above)
Comment (MyIgel) [reply lib/forcegraph.ts:111]: Its should be some kind of MapNode that has an `o` attribute with the actual node

**Review by weeman1337 — APPROVED** [review]
Tested, looks good and works!

**Pull request merged**